File: FFT/Lab2/OpenACC/c/cufftACC.c
#include "../../Inputs/c/fileReader.h"
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
// CUDA includes
#include <cufft.h>

//Input Size
#define SIZE 134217728

typedef double2 Complex;

//Host allocated pointers
Complex *h_inputFFT, *h_outputFFT, *h_outputInvFFT;

//Timing variables
struct timeval start, stop;
float execTime;

void allocate() {

    h_inputFFT = (Complex *) malloc(sizeof(Complex) * SIZE);
    h_outputFFT = (Complex *) malloc(sizeof(Complex) * SIZE);
    h_outputInvFFT = (Complex *) malloc(sizeof(Complex) * SIZE);

}
void initialize() {

    int i;
    double inputs[SIZE];
    
    //Loading real numbers from a fle
    char *fname = "../../Inputs/complexInputs.txt";
    readInputs(fname, inputs, SIZE);
    
    for (i=0; i<SIZE; ++i) {
        h_inputFFT[i].x = inputs[i];
        h_inputFFT[i].y = 0.0;
    }
}

void executeFFT() {

    int i;
    double error = 0.0;
    //Creating a plan for CUFFT
    cufftHandle plan;
    
    gettimeofday(&start, NULL);

    cufftPlan1d(&plan, SIZE, CUFFT_Z2Z, 1); // Z2Z is C2C with Double Precision
   
    #pragma acc host_data use_device(h_inputFFT, h_outputFFT, h_outputInvFFT)
    {
        //Performing FFT FORWARD
        cufftExecZ2Z(plan, (cufftDoubleComplex *) h_inputFFT,
                    (cufftDoubleComplex *) h_outputFFT,
                    CUFFT_FORWARD);
        //Performing FFT INVERSE
        cufftExecZ2Z(plan, (cufftDoubleComplex *) h_outputFFT,
                    (cufftDoubleComplex *) h_outputInvFFT,
                    CUFFT_INVERSE);
    }
    
    cufftDestroy(plan);

    //Transfering data back to the CPU for comparison
    #pragma acc update self(h_outputInvFFT[:SIZE]) wait
    
    gettimeofday(&stop, NULL);

    //Accumulating total error
    for (i=0; i<SIZE; ++i) {
        error += fabs(h_inputFFT[i].x - h_outputInvFFT[i].x/SIZE);
    }
    
    printf("\nMean Error = %e\n", error / SIZE);

    execTime = (float)(stop.tv_usec - start.tv_usec) / 1000000 + (float)(stop.tv_sec - start.tv_sec);
}

void deallocate() {

    free(h_inputFFT);
    free(h_outputFFT);
    free(h_outputInvFFT);
}

int main () {

    allocate();
    
    initialize();
    #pragma acc data copyin(h_inputFFT[:SIZE]), create(h_outputFFT[:SIZE], h_outputInvFFT[:SIZE])
    {
        executeFFT();
    }
    
    printf("\nOnly FFT:");
    printf("\nExecution Time: %.10f s\n\n", execTime);
    deallocate();

    return 0;
}

File: FFT/Lab2/OpenACC/fortran/cufftACC.f90
program MAIN
    
    use fileReaderF90
    use cufft

    implicit none
    integer, parameter :: SIGNAL_SIZE = 134217728
    double complex, allocatable, dimension(:) :: h_inputFFT, h_outputFFT, h_outputInvFFT
    real*8 :: execTime, start, finish
    
    allocate(h_inputFFT(SIGNAL_SIZE), h_outputFFT(SIGNAL_SIZE), h_outputInvFFT(SIGNAL_SIZE))
    
    call initialize() 
    
    !$acc data copyin(h_inputFFT) create(h_outputFFT, h_outputInvFFT)
    call executeFFT()
    !$acc end data
    
    print *, "Only FFT:"
    print *, "Execution Time:", execTime
    print *

    deallocate(h_inputFFT, h_outputFFT, h_outputInvFFT)





    contains
        subroutine initialize()
            implicit none
            integer :: i
            real*8, dimension(SIGNAL_SIZE) :: inputs
            
            ! Loading real numbers from a fle
            call readInputs("../../Inputs/complexInputs.txt", inputs, SIGNAL_SIZE);
            
            do i=1, SIGNAL_SIZE
                h_inputFFT(i) = dcmplx(inputs(i), 0.0)
            end do
        end subroutine initialize

        subroutine executeFFT()
            implicit none
            
            integer :: i, plan, err
            real*8 :: errorSum
            errorSum = 0.0
            
            call cpu_time(start)

            !Create a cufft plan for Z2Z i.e complex to complex in Double precision
            err = cufftPlan1d(plan, SIGNAL_SIZE, CUFFT_Z2Z, 1)

            !$acc host_data use_device(h_inputFFT, h_outputFFT, h_outputInvFFT)
            
            
            !Perform FFT out of place transform
            err = err + cufftExecZ2Z(plan, h_inputFFT, h_outputFFT, CUFFT_FORWARD)
            
            !Perform Inverse FFT in plance transform
            err = err + cufftExecZ2Z(plan, h_outputFFT, h_outputInvFFT, CUFFT_INVERSE)
            !$acc end host_data

            err = err + cufftDestroy(plan)

            !copy to host
            !$acc update self(h_outputInvFFT(:SIGNAL_SIZE)) wait
            
            call cpu_time(finish)
    
            !Accumulate total error
            do i=1, SIGNAL_SIZE
                errorSum  = errorSum + DABS(real(h_inputFFT(i)) - real(h_outputInvFFT(i))/SIGNAL_SIZE)
            end do
            
            print *
            print *, "Mean Error: ", errorSum / SIGNAL_SIZE
            print *

            execTime = finish - start
        end subroutine executeFFT
    
end program MAIN
